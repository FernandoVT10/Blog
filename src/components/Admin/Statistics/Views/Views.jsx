import Chart from "../Chart/Chart";
import Api from "../../../../ApiController";

import { useEffect, useState } from "react";

import "./Views.scss";

function Views() {
    const [views, setViews] = useState({
        total: "- -",
        month: "- -",
        day: "- -" 
    });
    const [chartType, setChartType] = useState("month");

    useEffect(() => {
        Api.get("views/getTotalViews/", true)
        .then(data => {
            if(data.month) {
                // add spaces in the number that separates the thousands

                const total = data.total.toString().replace(/\B(?=(\d{3})+(?!\d))/g, " ");
                const month = data.month.toString().replace(/\B(?=(\d{3})+(?!\d))/g, " ");
                const day = data.day.toString().replace(/\B(?=(\d{3})+(?!\d))/g, " ");

                setViews({ total, month, day });
            }
        });
    }, []);

    const totalViewsItemClass = chartType === "month" ? "active" : "";
    const monthViewsItemClass = chartType === "day" ? "active" : ""; 

    return (
        <div className="statistics-views">
            <div className="statistics-views__total-views">
                <div
                className={`statistics-views__total-views-item ${totalViewsItemClass}`}
                onClick={() => setChartType("month")}>
                    <span className="statistics-views__total-views-label">
                        Total Views
                    </span>
                    <span className="statistics-views__total-views-number">
                        { views.total }
                    </span>
                </div>
                <div
                className={`statistics-views__total-views-item ${monthViewsItemClass}`}
                onClick={() => setChartType("day")}>
                    <span className="statistics-views__total-views-label">
                        Month Views
                    </span>
                    <span className="statistics-views__total-views-number">
                        { views.month }
                    </span>
                </div>
                <div className="statistics-views__total-views-item disabled">
                    <span className="statistics-views__total-views-label">
                        Day Views
                    </span>
                    <span className="statistics-views__total-views-number">
                        { views.day }
                    </span>
                </div>
            </div>

            <div className="statistics-views__charts-container">
                <Chart type={chartType}/>
            </div>
        </div>
    );
}

export default Views;