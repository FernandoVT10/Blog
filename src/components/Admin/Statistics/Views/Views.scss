@use "../../../../styles/variables";

.statistics-views {
    background: variables.$primary;

    &__total-views {
        display: flex;
    }

    &__total-views-item {
        cursor: pointer;
        display: flex;
        padding: 15px;
        flex-direction: column;
        align-items: center;
        justify-content: center;
        color: white;
        transition: .3s;

        &:hover {
            background: variables.$black;
        }

        &.active {
            background: variables.$black;
        }

        &.disabled:hover {
            cursor: auto;
            background: transparent;
        }
    }

    &__total-views-number {
        font-weight: bold;
        font-size: 20px;
    }

    /*<--------------->*/
        /*CHARTS*/
    /*<--------------->*/

    &__charts-container {
        padding: 20px;
    }

    &__chart {
        width: 100%;
        height: 250px;
    }
}