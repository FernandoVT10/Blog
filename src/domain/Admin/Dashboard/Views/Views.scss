.dashboard-views {
    background: var(--primary);

    &__total-views {
        display: flex;
    }

    &__total-views-item {
        cursor: pointer;
        display: flex;
        padding: 15px;
        flex-direction: column;
        align-items: center;
        justify-content: center;
        color: var(--white);
        transition: .3s;

        &:hover {
            background: var(--black);
        }

        &.active {
            background: var(--black);
        }

        &.disabled:hover {
            cursor: auto;
            background: transparent;
        }
    }

    &__total-views-number {
        font-weight: bold;
        font-size: 20px;
    }

    /*<--------------->*/
        /*CHARTS*/
    /*<--------------->*/

    &__charts-container {
        padding: 20px;
    }

    &__chart {
        width: 100%;
        height: 250px;
    }
}