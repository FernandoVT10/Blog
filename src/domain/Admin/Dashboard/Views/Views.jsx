import Chart from "./Chart/Chart";

import ApiController from "../../../../services/ApiController";
import separateThousands from "../../../../services/separateThousands";

import { useEffect, useState } from "react";

import "./Views.scss";

export default () => {
    const [views, setViews] = useState({
        total: "- -",
        month: "- -",
        day: "- -" 
    });
    const [chartType, setChartType] = useState("month");

    useEffect(() => {
        ApiController.get("articles/views/getTotalViews/", true)
        .then(res => {
            if(res.data) {
                const { views } = res.data;

                const total = separateThousands(views.total);
                const month = separateThousands(views.month);
                const day = separateThousands(views.day);

                setViews({ total, month, day });
            }
        });
    }, []);

    const totalViewsItemClass = chartType === "month" ? "active" : "";
    const monthViewsItemClass = chartType === "day" ? "active" : ""; 

    return (
        <div className="dashboard-views">
            <div className="dashboard-views__total-views">
                <div
                className={`dashboard-views__total-views-item ${totalViewsItemClass}`}
                onClick={() => setChartType("month")}>
                    <span className="dashboard-views__total-views-label">
                        Total Views
                    </span>
                    <span className="dashboard-views__total-views-number">
                        { views.total }
                    </span>
                </div>
                <div
                className={`dashboard-views__total-views-item ${monthViewsItemClass}`}
                onClick={() => setChartType("day")}>
                    <span className="dashboard-views__total-views-label">
                        Month Views
                    </span>
                    <span className="dashboard-views__total-views-number">
                        { views.month }
                    </span>
                </div>
                <div className="dashboard-views__total-views-item disabled">
                    <span className="dashboard-views__total-views-label">
                        Day Views
                    </span>
                    <span className="dashboard-views__total-views-number">
                        { views.day }
                    </span>
                </div>
            </div>

            <div className="dashboard-views__charts-container">
                <Chart type={chartType}/>
            </div>
        </div>
    );
}