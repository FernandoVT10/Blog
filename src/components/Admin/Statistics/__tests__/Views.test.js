import { act } from "react-dom/test-utils";
import { render } from "react-dom";
import Views from "../Views/";

const MOCK_TOTAL_VIEWS = {
    total: 3319519,
    month: 22112,
    day: 5564
};

let container;

// Mock Chart JS

class ChartMock {
    static defaults = {
        global: {
            elements: {
                point: {}
            },
            legend: {}
        }
    };

    constructor(_, options) {
        this.data = options.data;
    }

    update() {}
}

window.Chart = ChartMock;

// Mock the canvas context

document.getElementById = () => ({ getContext: jest.fn() });

describe('<Articles/> Component', () => {
    beforeEach(() => {
        fetchMock.doMock();

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    describe("Check if renders correctly", () => {
        it('It should render without data', async () => {
            fetchMock
                .once(JSON.stringify([]))
                .once(JSON.stringify({}));

            await act(async () => {
                render(<Views/>, container);
            });
    
            expect(container).toMatchSnapshot();
        });

        it('It should render with data', async () => {
            fetchMock
                .once(JSON.stringify([]))
                .once(JSON.stringify(MOCK_TOTAL_VIEWS));

            await act(async () => {
                render(<Views/>, container);
            });
    
            expect(container).toMatchSnapshot();
        });
    });
});