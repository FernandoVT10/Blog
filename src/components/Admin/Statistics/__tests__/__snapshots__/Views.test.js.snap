// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Articles/> Component Check if renders correctly It should render with data 1`] = `
<div>
  <div
    class="statistics-views"
  >
    <div
      class="statistics-views__total-views"
    >
      <div
        class="statistics-views__total-views-item active"
      >
        <span
          class="statistics-views__total-views-label"
        >
          Total Views
        </span>
        <span
          class="statistics-views__total-views-number"
        >
          3 319 519
        </span>
      </div>
      <div
        class="statistics-views__total-views-item "
      >
        <span
          class="statistics-views__total-views-label"
        >
          Month Views
        </span>
        <span
          class="statistics-views__total-views-number"
        >
          22 112
        </span>
      </div>
      <div
        class="statistics-views__total-views-item disabled"
      >
        <span
          class="statistics-views__total-views-label"
        >
          Day Views
        </span>
        <span
          class="statistics-views__total-views-number"
        >
          5 564
        </span>
      </div>
    </div>
    <div
      class="statistics-views__charts-container"
    >
      <canvas
        class="statistics-views__chart"
        height="400"
        id="viewsChart"
        width="2400"
      />
    </div>
  </div>
</div>
`;

exports[`<Articles/> Component Check if renders correctly It should render without data 1`] = `
<div>
  <div
    class="statistics-views"
  >
    <div
      class="statistics-views__total-views"
    >
      <div
        class="statistics-views__total-views-item active"
      >
        <span
          class="statistics-views__total-views-label"
        >
          Total Views
        </span>
        <span
          class="statistics-views__total-views-number"
        >
          - -
        </span>
      </div>
      <div
        class="statistics-views__total-views-item "
      >
        <span
          class="statistics-views__total-views-label"
        >
          Month Views
        </span>
        <span
          class="statistics-views__total-views-number"
        >
          - -
        </span>
      </div>
      <div
        class="statistics-views__total-views-item disabled"
      >
        <span
          class="statistics-views__total-views-label"
        >
          Day Views
        </span>
        <span
          class="statistics-views__total-views-number"
        >
          - -
        </span>
      </div>
    </div>
    <div
      class="statistics-views__charts-container"
    >
      <canvas
        class="statistics-views__chart"
        height="400"
        id="viewsChart"
        width="2400"
      />
    </div>
  </div>
</div>
`;
